### PR TITLE
Add tooling and restructure project as a result

### PR DESCRIPTION
**Pull request by tdeekens**
> This pull request aims to add some tooling to make this project easier to work with. It's aligned with the Node.JS SDK and flopflip.

- Adds CircleCI Integration
- Adds lerna maintaining this as a monorepo
- Adds Jest for testing
- Adds ESLint for code linting
- Fixes errors found in the process
- Adds an example test for the API Extension examples
- Adds Renovate (auto updates or dependencies)

### CODE REVIEW

**Comment by tdeekens** [comment]
Weeeh. Tests are running both on Node.JS 6 and 8 with automatic code coverage reports @cneijenhuis!

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# Remove whitespace characters preceding newline characters
+trim_trailing_whitespace = true
+
+# Two space indentation for JavaScript files
+[*.{js,json}]
+indent_size = 2
+
+# Disable trimming trailing whitespaces so that double space newlines work
+[*.md]
+trim_trailing_whitespace = false
```
Comment (adnasa) [inline .editorconfig]: out of curiosity; Does prettier respect this rule?

---

isn't it simpler to read the intention of the author  by just using/forcing the use of `<br/>`?

`  ` is hidden and hard to spot when reviewing/reading/updating docs
with `<br />`, you at least see the intent and the cognitive overload is out of the way

File: lerna.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    },
+    "cacheDir": ".changelog",
+    "ignoreCommitters": ["renovate", "renovate-bot"]
+  },
+  "command": {
+    "publish": {
+      "message": "chore(release): releasing package"
+    }
+  }
+}
```
Comment (tdeekens) [inline lerna.json]: lerna is a tool to manage monorepos. It looks through everything in `packages` for a `package.json` and installs things and does a bunch more.

File: packages/example-inventory/001-hello-world/index.js [hunk 3]
```diff
@@ -11,7 +11,8 @@ exports.helloWorld = function helloWorld(req, res) {
     res.status(400).send('No sku defined!');
   } else {
     // Everything is okay.
+    // eslint-disable-next-line no-console
     console.log(req.body.sku);
-    res.status(200).send('Inventory: ' + parseInt(Math.random() * 1000));
+    res.status(200).send(`Inventory: ${parseInt(Math.random() * 1000, 10)}`);
```
Comment (tdeekens) [inline packages/example-inventory/001-hello-world/index.js]: This was an ESLint warning. `parseInt` should always have a radix!

File: .renovaterc.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "extends": [
+    "config:base",
+    ":pinOnlyDevDependencies",
+    "group:monorepos",
+    "schedule:weekly"
+  ]
+}
```
Comment (adnasa) [inline .renovaterc.json]: 👍

File: README.md [hunk 5]
```diff
@@ -1,5 +1,19 @@
...
+</p>
+
+<p align="center">
+  <a href="https://circleci.com/gh/commercetools/commercetools-serverless-examples">
+    <img alt="CircleCI Status" src="https://circleci.com/gh/commercetools/commercetools-serverless-examples.svg?style=shield&circle-token=aa">
+  </a>
+  <a href="https://codecov.io/gh/commercetools/commercetools-serverless-examples">
+    <img alt="Codecov Coverage Status" src="https://img.shields.io/codecov/c/github/commercetools/commercetools-serverless-examples.svg?style=flat-square">
+  </a>
+  <img alt="Made with Coffee" src="https://img.shields.io/badge/made%20with-%E2%98%95%EF%B8%8F%20coffee-yellow.svg">
```
Comment (adnasa) [inline README.md]: 🇸🇪 coffee, I might add

File: packages/example-inventory/002-connecting-ct/index.js [hunk 6]
```diff
@@ -58,22 +62,31 @@ exports.helloWorld = function helloWorld(req, res) {
...
+            .json({ error: 'Inventory for SKU not found!' });
+        }
+
+        return res.status(200).json(results);
+      })
+      .catch(e => {
+        // eslint-disable-next-line no-console
+        console.error(e);
+        res.status(400).json(e);
+      });
```
Comment (tdeekens) [inline packages/example-inventory/002-connecting-ct/index.js]: All changes are due to formatting by prettier. No more hand formatting! Rules are enforced through ESLint. Just run `npm run format` or setup an editor integration for format on save. It's like gofmt.

File: packages/extension-base/ctp-extension-azure-adapter.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
+// Azure
+
+const createReject = ctx => error => {
+  ctx.res = {
+    status: 400,
+    body: {
+      errors: Array.isArray(error) ? error : [error],
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-azure-adapter.js]: Inlined the check.

File: packages/extension-base/ctp-extension-azure-adapter.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  ctx.done();
+};
+
+const ctpResponse = ctx => ({
+  reject: createReject(ctx),
+  accept: createAccept(ctx),
+});
+
+module.exports = {
+  create: fn => (ctx, req) => {
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-azure-adapter.js]: Slight API change for a `adapter.create(fn)(ctx, res)`

File: packages/extension-base/ctp-extension-gcf-adapter.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+const createReject = ctx => error => {
+  ctx.status(400).json({
+    errors: Array.isArray(error) ? error : [error],
+  });
+};
+
+const createAccept = ctx => (update = []) => {
+  ctx.status(200).json({
+    actions: Array.isArray(update) ? update : [update],
+  });
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-gcf-adapter.js]: I like GCF's API so much more.

File: packages/extension-base/ctp-extension-gcf-adapter.spec.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    it('should set status to `200`', () => {
+      expect(context.status).toHaveBeenCalledWith(400);
+    });
+
+    it('should pass errors to response', () => {
+      expect(response.json).toHaveBeenCalledWith({ errors });
+    });
+  });
+});
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-gcf-adapter.spec.js]: Just giving some examples on how to test things with Jest.

**Comment by tdeekens** [comment]
Ready to merge now.

File: .circleci/config.yml [hunk 11]
```diff
@@ -0,0 +1,92 @@
+node6Environment: &node6Environment
+  docker:
+    - image: circleci/node:6
+  working_directory: ~/nodejs
+node8Environment: &node8Environment
+  docker:
+    - image: circleci/node:8
+  working_directory: ~/nodejs
```
[the remark this answers is not in the document]
Comment (cneijenhuis) [reply .circleci/config.yml]: noob question: Why both?

Does node 8 deprecate/remove/change some things from node 6? I kinda expected it to be backwards-compatible.

File: packages/extension-base/ctp-extension-azure-adapter.js (hunk 7 above)
Comment (cneijenhuis) [reply packages/extension-base/ctp-extension-azure-adapter.js]: thanks 👍

File: packages/extension-base/ctp-extension-azure-adapter.js (hunk 8 above)
Comment (cneijenhuis) [reply packages/extension-base/ctp-extension-azure-adapter.js]: I do expect to *eventually* create an adapter for subscriptions as well.

I'm fine with using `create` in the function name, but I'm wondering if it isn't better to be more explicit (e.g. `createExtensionAdapter` and `createSubscriptionAdapter`).

File: packages/extension-base/ctp-extension-gcf-adapter.js (hunk 9 above)
Comment (cneijenhuis) [reply packages/extension-base/ctp-extension-gcf-adapter.js]: Agreed. From what I get, this is the `express.js` API.

File: packages/extension-base/ctp-extension-gcf-adapter.spec.js (hunk 10 above)
Comment (cneijenhuis) [reply packages/extension-base/ctp-extension-gcf-adapter.spec.js]: 🏆 🙇 🥇

File: .circleci/config.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (tdeekens) [reply .circleci/config.yml]: No hard opinion. Its the config of the JS SDK. Felt like when cloud function support two versions we should test those too. Especiall when we do not transpile.

File: packages/extension-base/ctp-extension-azure-adapter.js (hunk 8 above)
Comment (tdeekens) [reply packages/extension-base/ctp-extension-azure-adapter.js]: Ah yeah sure. Whatever you feel is more sustainable.

File: .editorconfig (hunk 1 above)
Comment (tdeekens) [reply .editorconfig]: I think prettier does not touch this as the white space would be between words. Maybe a br would be easier to read but isnt possible in all places in md files.

File: packages/extension-base/ctp-extension-gcf-adapter.js (hunk 9 above)
Comment (tdeekens) [reply packages/extension-base/ctp-extension-gcf-adapter.js]: Yeah its Hapi/Express like. Maybe even more Hapi than Express. The Azure is more barebones Node Http.

**Comment by tdeekens** [comment]
Glad it helps. Ill then change the create name to be namespaced tomorrow and merge this in.

File: packages/extension-base/ctp-extension-azure-adapters.spec.js [hunk 12]
```diff
@@ -1,4 +1,4 @@
-const adapters = require('./ctp-extension-azure-adapters');
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-azure-adapters.spec.js]: Did this in two steps. I think we export multiple adapters if I understand your vision correctly.

File: packages/extension-base/ctp-extension-azure-adapters.spec.js [hunk 13]
```diff
@@ -1,4 +1,4 @@
...
+const { createExtensionAdapter } = require('./ctp-extension-azure-adapters');
```
Comment (tdeekens) [inline packages/extension-base/ctp-extension-azure-adapters.spec.js]: ...which now gives us the "power" to just destructure the import.

**Pull request merged**